ListStore integration in PostStore

oguzkocer opened the pull request:
This PR implements the `ListStore` actions in `PostStore` with some other minor improvements.

### ListStore Changes

* I've updated the `ListStore` to take the fetch list function as parameter. This way we don't have to update the ListStore for different list descriptors and it plays much better with other modules. (Thanks @AmandaRiu for pointing this issue out to me)
* `PostListDescriptor` now has a parameter for `pageSize`. This was mostly done to make it easier to write connected tests as having a requirement for a site with more than 100 posts didn't really make sense to me. This means the client could potentially change the `pageSize` in each fetch, but if they want to do it, more power to them. `ListStore` should be able to handle that pretty comfortably.

### PostStore implementation of ListStore

This implementation is fairly straightforward. One thing to note is that I am not actually reusing the `FETCH_POSTS` action as the behavior we want is different and that action is still used elsewhere. Ideally, we'll just remove that action, as well as `FETCH_PAGES` once we move everything to `ListStore`. For now, I think the cleanest solution is to keep them separate.

`PostStore` will also let `ListStore` know about the changes to `PostModel`s such as updates and removes so it can update the lists. Observing for the `OnListChanged` and `OnListItemsChanged` is enough to get these updates/removes in the client for "free".

### Not Implemented

When a post's status, date etc changes, the lists will not be updated. That means, if the user trashes a post, it'll not immediately removed from the list for `published` posts. The list will need to be refreshed for that change to effect in the client.

This behavior is not different from the current clients, so we are not losing anything, on the contrary refreshing a list is a very inexpensive and fast operation, so it'll work much better. I have some thoughts about how we might be able to achieve a better behavior around this (using ListState.NEEDS_REFRESH), but I don't have any time to work on this, so there won't be any changes to this at least for v1.

@malinajirka You previously mentioned that you prefer to have the tests in the PR even if it makes it longer, so I am keeping the connected tests here, but if you want, I can definitely separate it into another PR. Also, while you're reviewing, if you want to play with the test UI, you can revert `94a7f81508f5f8b032f87f7040eeb9098599ee98` which will add a new page to the example app where all this is implemented (in a non-polished way). I removed those changes to make it easier to review this PR.

malinajirka commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/PostStore.java:
```diff
@@ -361,6 +433,50 @@ public PostModel getPostByRemotePostId(long remoteId, SiteModel site) {
+                return Collections.emptyList();
+            }
+        }
+        String orderBy;
```
We might consider removing the `default` branch and setting the orderBy to null here. I know we have a lint rule which enforces providing all the cases, however the default branch feels like a dead code. Moreover, by omitting the default branch it's explicit that there are no other cases + if someone introduces a new enum value, they won't be even able to build the code. Wdyt?

oguzkocer replied:
I agree with this 100%. I only left the default so the AS wouldn't complain to me, but AS is still complaining so no reason not to change it. At least this works better with Kotlin.

Fixed in 5bc7f8c5146d916f527cb6ba62f505eb8e1fdabc.

malinajirka commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/PostListConnectedTestHelper.kt:
```diff
@@ -0,0 +1,154 @@
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+class PostListConnectedTestHelper(
```
General comment: We might consider using assertXYZ(message, A,B) rather than a comment and assertXYZ(A,B). Benefit is that it's easier to see why the test failed in Travis.

oguzkocer replied:
Fixed in 5b8d12c24490df16f71fb479274779f52cd9f8fe.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/post/PostRestClient.java:
```diff
@@ -91,26 +95,52 @@ public void onErrorResponse(@NonNull WPComGsonNetworkError error) {
 
-        Map<String, String> params = new HashMap<>();
+        String fields = "ID,modified";
+        Map<String, String> params = getFetchPostListParameters(false, offset, listDescriptor.getPageSize(),
```
Fixed in dd0ce1a6aa06208c52126b2539741cec6629b8ce.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/post/PostRestClient.java:
```diff
@@ -91,26 +95,52 @@ public void onErrorResponse(@NonNull WPComGsonNetworkError error) {
+        String url = WPCOMREST.sites.site(listDescriptor.getSite().getSiteId()).posts.getUrlV1_1();
 
-        Map<String, String> params = new HashMap<>();
+        String fields = "ID,modified";
```
Fixed in dd0ce1a6aa06208c52126b2539741cec6629b8ce.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/post/PostXMLRPCClient.java:
```diff
@@ -110,28 +114,65 @@ public void onErrorResponse(@NonNull BaseNetworkError error) {
+        fields.add("post_id");
+        fields.add("post_modified");
+        List<Object> params =
+                getFetchPostListParameters(site.getSelfHostedSiteId(), site.getUsername(), site.getPassword(), false,
```
Fixed in dd0ce1a6aa06208c52126b2539741cec6629b8ce.

oguzkocer commented:
@malinajirka Thanks a lot for the review. I have addressed your comments and left replies for each commit. Let me know if there is anything else we should improve.

malinajirka commented:
Thank you @oguzkocer for addressing all my comments!! 

LGTM
